### DISCUSSION TITLE
Is there a plan for interactive walk/bike directions (or is it possible already)?

### DISCUSSION THREAD

**Discussion by andrewtavis** [maplibre/maplibre-gl-directions#228] [General]
Hi there :) First and foremost, this plugin has been such a joy to work with and really takes the MapLibre experience to the next level 😊 Thanks so much for it!

The team at [activist-org](https://github.com/activist-org/activist) has been using `maplibre-gl-directions` in the development of our MVP, and one thing that's come up is wanting interactive walking/biking directions. We made [activist#685](https://github.com/activist-org/activist/issues/685) to explore and implement walking directions as the default, as it's obviously very important for us to offer this as a means of people getting to demonstrations and the like. A contributor and I were playing around with it, and then I realized that in the [profiles demo](https://maplibre.org/maplibre-gl-directions/#/examples/multiple-profiles) it says `Note that interactivity is not supported for multiple profiles`. The code that we're playing around with is [here](https://github.com/activist-org/activist/blob/91a230a661faf81af634328d2dbc32d014d9c7fc/frontend/components/media/MediaMap.vue#L140), and a view of it on a demo event page can be seen on [activist.org/events/123/about](https://activist.org/events/123/about) (any integer pre MVP loads a playground event).

Am I right in assuming that walking and biking are not supported for the interactive directions at this time? It looks like we're doing it correctly, but maybe we're missing something 🤔 I'm asking, because specifically the walk/bike directions are not able to go through the park in the linked demo event page, whereas in the [OSRM demo](https://map.project-osrm.org/?z=16&center=52.512643%2C13.358774&loc=52.510008%2C13.364353&loc=52.513952%2C13.357229&hl=en&alt=0&srv=2) you are able to set both walking and biking directions that go over park paths. If it's not yet implemented, is it on the roadmap?

Thanks again so much!

**Comment by smellyshovel** [comment]
@andrewtavis Hello, Andrew.

First of all, thanks for the kind words. I appreciate it and I'm glad my code is of use for someone!

Currently, as I can remember, if you try to use multiple profiles together with interactivity enabled, you won't get any exception/error or anything else that would stop you from doing that. However, if you still do it and encounter some errors or unexpected behavior, we won't react to issues opened about it, since *it is not supported*. So, this is what we mean, when we say that we don't support multiple directions with interactivity enabled. Not that it definitely won't work.

The possibility of having interaction enabled for multi-profiled requests wasn't tested *enough*. We surely tried to support interactivity for them, but most probably there were subtle bugs that required a lot of attention. And given that we personally didn't need the interactivity together with multiple profiles support for our use cases, we haven't made a lot of effort to try to make them work together properly.

Of the top of my head, I can come up with at least the following questions right away (if I think of proper interactivity support for multiple profiles requests):
1. When adding a new waypoint (with a mouse click), which profile should it be of?
2. When adding the first waypoint, which profile should it be?
3. When dragging the line between two waypoints, the newly added one should be of which profile?

What's for me personally, I can't answer them right away. At least some kind of research, planning and thinking is required. Though, I guess that implementing a control to pre-select profile for new waypoints would solve all the issues. On the other hand, what about the cases when users don't want to use the built-in control?

In other words, there should be sane defaults, if we want to support that. And establishing them requires at least some discussion with the users of the plugin.

Does it mean that it's not planned to support such a functionality? Definitely, no.
Does it mean though that it's on my personal roadmap? As well, no. At least for the near future.
In other words, a PR would be _very much_ appreciated! ~~There may even be nothing to do really, in case everything already works properly. It's just needed to make sure it really does. And to document the existing behavior.~~

**Comment by andrewtavis** [comment]
Hey @smellyshovel 👋 Thanks so much for your quick reply!

I guess I'm getting a bit confused with things 😇 I just assumed that `Note that interactivity is not supported for multiple profiles` was maybe the cause of us not being able to get directions for walking or biking, but now what I see is that that's referring specifically to a situation where a person would want to start with a car and then walk the rest of the way and switch the directions as they go. I read `multiple profiles` like "Interactivity isn't working for some profiles". An interesting problem 🤔 I'll answer your questions, and will try to explain a bit more what I mean for our use case afterwards:

1. I think this would potentially require a menu option or something like this, right?
  - Say you currently are on drive, then you put a first and second waypoint for driving from A to B, and then you click the menu option like in [the designs we have](https://www.figma.com/file/I9McFfaLu1RiiWp5IP3YjE/activist_public_designs?type=design&node-id=2673%3A20111&mode=design&t=vm4LpPlPqtMaIuOG-1) and select walk such that any next waypoint after that is walking
  - Adding in a right click option for it such that on right click an option to select which profile to use for the point that you're placing would be optimal, but also much more difficult
2. The user should set a default, with it being drive from the start unless they change it
3. I'd say the route would default to the last waypoint in a path, but then also not sure :)

Back to what we're dealing with, which it sounds like there is a solution, how do we set it to do walking or biking directions? The code from our [MediaMap component](https://github.com/activist-org/activist/blob/91a230a661faf81af634328d2dbc32d014d9c7fc/frontend/components/media/MediaMap.vue#L140) is the following:

```ts
const directions = new MapLibreGlDirections(map, {
  api: "https://router.project-osrm.org/route/v1",
  profile: "foot",
  // profile: "bike",
  // profile: "driving",
  requestOptions: {
    alternatives: "true",
  },
  layers,
});

directions.interactive = true;
```

With the above, if we try to set a new marker somewhere in a park, then the directions go along the streets and then there's a perpendicular dashed line to the point within the park ([see implementation live](https://activist.org/en/events/123/about)). This is the same for if we comment `"foot"` and uncomment `"bike"`, which for me is indicating that we're not actually getting walking or biking directions as the [OSRM demo](https://map.project-osrm.org/?z=16&center=52.512643%2C13.358774&loc=52.510008%2C13.364353&loc=52.513952%2C13.357229&hl=en&alt=0&srv=2) allows you to take park paths.

Any thoughts on the above use case for us? Would also be happy to continue the discussion of how the multiple profile routes could work!

Thanks again!

**Comment by smellyshovel** [comment]
@andrewtavis 

> I guess I'm getting a bit confused with things 😇 I just assumed that Note that interactivity is not supported for multiple profiles was maybe the cause of us not being able to get directions for walking or biking, but now what I see is that that's referring specifically to a situation where a person would want to start with a car and then walk the rest of the way and switch the directions as they go.

You're absolutely right!

> I read multiple profiles like "Interactivity isn't working for some profiles".

I'm sorry for the confusion. Are there maybe any ideas on how I could improve the wording in README and Demo pages?

---

The code you provided looks good to me. I think I know what's the deal here.

Seems to me like OSRM changed the way they work with profiles.

Here's an example request URL when I'm using the OSRM demo site: `https://routing.openstreetmap.de/routed-bike/route/v1/driving/11.744384765625002,49.1314084139986;11.981775234375009,50.80428797681432?overview=false&alternatives=true&steps=true`. When I change the profile to "foot", then the URL looks like this: `https://routing.openstreetmap.de/routed-foot/route/v1/driving/11.744384765625002,49.1314084139986;11.981775234375009,50.80428797681432?overview=false&alternatives=true&steps=true`.

You can see that the part starting with `driving` remains the same. However, that was not the case when I created the first implementation of the plugin.

For your particular case, I'd suggest you try the following code:

```
const directions = new MapLibreGlDirections(map, {
  api: "https://routing.openstreetmap.de/routed-bike/route/v1/",
  profile: "driving",
  requestOptions: {
    alternatives: "true",
  },
  layers,
});

directions.interactive = true;
```

This should work for you, if I'm not mistaken. However, in order to change the profile to e.g. "foot", you'd have to update the URL, without touching the `profile` field.

Why it works the way it works now? I guess because when OSRM is provided with an unknown for it profile, then it defaults to driving. And it's "unknown" to it (I guess) because they changed the way the URLs are built.

However, if they really did change the part of the spec regarding the request URLs, then it means that the plugin is outdated and I need to fix the way URLs are built. For that, I'd ask you to create an issue here in the repo.

**Comment by smellyshovel** [comment]
On the other hand, the spec shows the old-fashioned URLs...

![Screenshot_20240422_132836](https://github.com/maplibre/maplibre-gl-directions/assets/9782236/d7c07c4e-ae36-4252-898f-f1c79af566a1)

TBH, I'm not sure why URLs they use on the Demo page look different from what's described in the spec. I encourage you to try the approach I suggested, and if it fixes the issue, then we think who's right and who's wrong and whether there's a need to fix something on our end.

Have a nice day!

**Comment by andrewtavis** [comment]
Thanks for the further replies and all the details, @smellyshovel! I'll give the fix a try tonight and report back on how it goes 😊

**Comment by andrewtavis** [comment]
I was too excited to not try it, @smellyshovel, and it's working! 🎉 So great 😊

I'm going to try out some other routing options, but it looks like the API is the main change here. The only difference between `profile: "driving"` and the others while using `https://routing.openstreetmap.de/routed-bike/route/v1/` is the color of the route from the default purple to teal for biking and blue for walking.

I'll try out a few more routing apis and try to set it up so we have one for walking and one for driving that are OpenStreetMap based. With that we'll be ready to start working on map options to change the routing options 🥳

To your question:
- I'd say that `Note that interactivity is not supported for multiple profiles in a single route` would have been enough for me to understand it better from the start :)

Do you see any further steps needed here?

Thanks again so much!

**Comment by smellyshovel** [comment]
> it's working! 🎉 So great 😊

Ok, glad to hear I could help.

> is the color of the route from the default purple to teal for biking and blue for walking.

Yeah, expected. Those colors depend on the `profile` setting, and now the `profile` setting is hard-coded in your case (the base URL changes instead), so no wonder the colors don't change when you only change the `api` and not the `profile`. 

On the other hand, you're still free to re-style it all to your liking as it's done [here](https://maplibre.org/maplibre-gl-directions/#/examples/restyling).

> I'd say that Note that interactivity is not supported for multiple profiles in a single route would have been enough for me to understand it better from the start :)

Ok, glad to head that, thanks for the feedback.

> Do you see any further steps needed here?

No, I don't think so. There definitely were some changes on the level of OSRM, but I'm not sure whether I need to fix something on my end or not. Let's see if there'd be more similar issues from other users, as it would indicate that something in the library really works in an unexpected way. But for now, we're all set up, I guess.

> Thanks again so much!

Thanks for using the plugin! Don't hesitate to ask me in case you need some guidance.
